test/validation/api/ipsec/ipsec_test_in.c

```diff
@@ -1741,6 +1742,411 @@ static void test_ipsec_sa_print(void)
 	ipsec_sa_destroy(in_sa);
 }
 
+static void test_in_ipv4_esp_reass_success_two_frags(odp_ipsec_sa_t out_sa,
+						     odp_ipsec_sa_t in_sa)
+{
+	ipsec_test_part test_in_1[2] = {
```

Comment:
@JannePeltonen I was wondering if we could make one ipsec_test_part hold multiple packets and use num_pkt to indicate the actual number of packets to be acted upon.

We will need to add this handling in ipsec_check_out_in_one function. And rename it to ipsec_check_out_in or ipsec_check_out_in_multi() or something like that. This would be a broader change, so I wanted to first float this idea first.

Reply:
I think ipsec_check_out_in_one() and its usage has gotten increasingly messy and this PR is making things even worse (passing invalid inbound SA to change the behaviour etc). It is partly because the function tries to do two things, outbound and inbound testing, at once.

I think it would be better to get rid of ipsec_check_out_in_one(), and maybe also get rid of ipsec_check_out_one() and ipsec_check_in_one(), and have single function for doing outbound or inbound IPsec processing and the associated checking. Then ipsec_test_part would not need to contain both inbound and outbound info like it does now in a bit messy way but could concentrate in just one (inbound or outbound, should not really matter) IPsec operation. Combined outbound + inbound tests would be handled by the caller (which could still be a wrapper and not necessarily the top level test case code) by calling the new function twice (for outbound and then for inbound, using the packet it got from the outbound test).

So, I would propose making a cleanup patch which would sort out the ipsec_check_out_in_one() mess and then rebasing the reassembly validation pathces on top of that. What do you think?

Reply:
> So, I would propose making a cleanup patch which would sort out the ipsec_check_out_in_one() mess and then rebasing
> the reassembly validation pathces on top of that. What do you think?

I had similar thoughts while preparing the patches for submission. So thought of giving you an early view so that we can align the efforts. I agree on your suggestion that we should redesign the functions to make it do one task and then let caller take care of running combinations. But since it would take time, do you think we can merge the API change first and take up the cleanup & reassembly UT as separate PRs?

Reply:
> do you think we can merge the API change first and take up the cleanup & reassembly UT as separate PRs?

I am not sure about that. Maybe I could try writing the cleanup patch if that would help.

Reply:
@JannePeltonen That should be fine. Thanks!